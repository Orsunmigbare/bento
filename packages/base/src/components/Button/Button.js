/**
 * @class Button
 */
import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"
import {
  targetSize,
  spacing,
  borderRadius,
  font,
  color,
  themify
} from "@staccx/theme"
const tinycolor = require("tinycolor2")

export const BUTTON = "Button"
export const ButtonStyle = css`
  background-color: ${color("primary")};
  color: ${p =>
    tinycolor
      .mostReadable(color("primary")(p), ["#fff"], {
        includeFallbackColors: true,
        level: "AA",
        size: "large"
      })
      .toString()};
  border-width: 0;
  border-radius: ${borderRadius()};
  padding: ${spacing("tiny")} ${spacing("medium")};
  cursor: pointer;
  min-height: ${targetSize()};
  text-decoration: none;
  text-align: center;
  display: inline-block;
  font-size: ${font("base")};
  font-family: ${font("body", "type")};
  line-height: 32px;
  font-weight: 600;
  white-space: nowrap;
  justify-content: center;
  align-items: center;
  margin-right: ${spacing("small")};
  margin-bottom: ${spacing("small")};
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  &:hover,
  &:focus {
    outline: none;
    background-color: ${p =>
      tinycolor(color("primary")(p))
        .darken(5)
        .toString()};
    color: ${p =>
      tinycolor
        .mostReadable(tinycolor(color("primary")(p)).darken(5), ["#fff"], {
          includeFallbackColors: true,
          level: "AA",
          size: "large"
        })
        .toString()};
  }

  &:last-child,
  &:last-of-type {
    margin-right: 0;
  }

  &:last-child:first-child {
    margin-bottom: 0;
  }

  ${themify(BUTTON)};
`
const Button = styled.button`
  ${ButtonStyle};
`

const defaultProps = {
  onClick: null,
  to: "#",
  className: ""
}

const propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element,
    PropTypes.string,
    PropTypes.array
  ]).isRequired,
  onClick: PropTypes.func,
  className: PropTypes.string,
  to: PropTypes.string
}

Button.defaultProps = defaultProps
Button.propTypes = propTypes

/** @component */
export default Button
