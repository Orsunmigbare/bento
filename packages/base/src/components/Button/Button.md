Button

```jsx harmony
<Button>Tester</Button>
```
