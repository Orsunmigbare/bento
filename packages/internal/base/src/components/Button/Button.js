import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"
import {
  targetSize,
  spacing,
  borderRadius,
  fontFamily,
  fontWeight,
  font,
  color,
  themify
} from "@staccx/theme"
import themePropTypes from "../constants/themePropTypes"
const tinycolor = require("tinycolor2")

const tProps = {
  button: {
    name: "Button",
    description: "Button styles",
    type: themePropTypes.style
  }
}

const ButtonComp = styled.button`
  background-color: ${color.primary};
  color: ${p =>
    tinycolor
      .mostReadable(color.primary()(p), ["#fff"], {
        includeFallbackColors: true,
        level: "AA",
        size: "large"
      })
      .toString()};
  border-width: 0;
  border-radius: ${borderRadius};
  padding: ${spacing.tiny} ${spacing.medium};
  cursor: pointer;
  min-height: ${targetSize.normal};
  text-decoration: none;
  text-align: center;
  display: inline-block;
  font-size: ${font.base};
  font-family: ${fontFamily.body};
  line-height: 32px;
  font-weight: ${fontWeight.bold};
  white-space: nowrap;
  justify-content: center;
  align-items: center;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  transition: background 0.2s ease;

  &:hover,
  &:focus {
    outline: none;
    background-color: ${p =>
      tinycolor(color.primary()(p))
        .darken(5)
        .toString()};
    color: ${p =>
      tinycolor
        .mostReadable(tinycolor(color.primary()(p)).darken(5), ["#fff"], {
          includeFallbackColors: true,
          level: "AA",
          size: "large"
        })
        .toString()};
  }

  ${themify(tProps.button.name)};
`
/**
 * Button component. Use to click stuff
 */
const Button = props => <ButtonComp {...props} />

const defaultProps = {
  onClick: null,
  to: "#",
  className: ""
}

const propTypes = {
  /**
   * The children of the lamb
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element,
    PropTypes.string,
    PropTypes.number,
    PropTypes.array
  ]).isRequired,
  onClick: PropTypes.func,
  className: PropTypes.string,
  to: PropTypes.string,
  test: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}

Button.defaultProps = defaultProps
Button.propTypes = propTypes
Button.themeProps = tProps

export default Button
