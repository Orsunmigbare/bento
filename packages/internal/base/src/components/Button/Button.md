Button

```jsx harmony
<div>
  <Button>Tester</Button>
</div>
```
