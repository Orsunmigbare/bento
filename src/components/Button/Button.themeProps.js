import { themePropTypes } from "../../theming"

const themeProps = {
  button: {
    name: "Button",
    description: "Button styles",
    type: themePropTypes.style
  }
}
export default themeProps
