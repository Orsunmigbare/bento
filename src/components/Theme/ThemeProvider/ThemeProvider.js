import React from "react"
import { ThemeProvider as StyledThemeProvider } from "styled-components"
import PropTypes from "prop-types"
import WebFonts from "../WebFonts"
import GlobalStyle from "../GlobalStyle"
import { theme as instance } from "../../../theming"
import { setLogLevel } from "../../../theming/theme.logger"
import { normalizeLevel } from "../../../utils/loglevelUtils"

/**
 * ThemeProvider is used to wrap you app. It will provide children with correct context
 * It also ensures usage of GlobalStyle and WebFonts
 * @param children
 * @param theme (optional)
 * @return {*}
 * @constructor
 */
const ThemeProvider = ({ children, theme, level }) => {
  React.useEffect(() => {
    // NOTE: This sets loglevel for THEME and not COMPONENTS.
    setLogLevel(normalizeLevel(level))
  }, [level])
  return (
    <StyledThemeProvider theme={theme}>
      <>
        <GlobalStyle />
        <WebFonts />
        {children}
      </>
    </StyledThemeProvider>
  )
}

ThemeProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element,
    PropTypes.func,
    PropTypes.string
  ]).isRequired,
  /**
   * Theme to override instance
   */
  theme: PropTypes.object
}
ThemeProvider.defaultProps = {
  theme: instance
}

export default ThemeProvider
