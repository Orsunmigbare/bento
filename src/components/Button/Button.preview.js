import React from "react"
import Button from "./Button"

const preview = {
  title: "Button",
  category: "Components/Buttons",
  tags: ["component", "clickable", "button"],
  component: Button,
  render: props => <Button {...props} />
}

export default preview
