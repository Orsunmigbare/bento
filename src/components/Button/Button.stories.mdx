import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import Button from "./Button"

<Meta title="Components|Button" component={Button}/>


# Button
Forklaring p√• hva det er og hvordan den kan brukes


# import

```javascript
Import {Button} from "@staccx/bento"
```

# Types of buttons


## Standard Button
This is the standard button



<Preview>
  <Story name="Standard">
   <Button>Knapp</Button>
  </Story>
</Preview>


## Two buttons
check console for onClick event

<Preview>
  <Story name="Two Buttons">
  <>
   <Button>Knapp</Button>
   <Button>Knapp</Button>
   </>
  </Story>
</Preview>

## Long String
<Preview>
  <Story name="Long String">
   <Button>KnappKnappKnappKnappKnappKnappKnappKnappKnappKnappKnappKnappKnappKnappKnapp</Button>
  </Story>
</Preview>



# Props for button
<Props of={Button} />
