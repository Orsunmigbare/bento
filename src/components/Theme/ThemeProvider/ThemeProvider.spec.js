import renderer from "react-test-renderer"
import React from "react"
import ThemeProvider from "./ThemeProvider"
import Button from "../../Button/Button"

describe("Theme Provider", () => {
  describe("Button", () => {
    it("Button", () => {
      const tree = renderer
        .create(
          <ThemeProvider>
            <Button>Click me</Button>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
