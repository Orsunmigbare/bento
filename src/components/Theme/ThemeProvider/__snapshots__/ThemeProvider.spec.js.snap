// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme Provider Button Button 1`] = `
<button
  className="sc-AykKC cAIyoR"
  onClick={null}
>
  Click me
</button>
`;
